Fix detection of tuple with TRest

Umdrahschua opened the pull request:
I run in an issue during getting all properties for `org.freedesktop.timesync1.Manager`.

During debugging I was able to find the code what is not working correct. The code cannot detect a tuple with only one element, because `typeInfo.IsLayoutSequential` returns `true`.


See also issue #215 and [compare branches main <-> test/timesysc1](https://github.com/tmds/Tmds.DBus/compare/main...Umdrahschua:Tmds.DBus:test/timesync1)

tmds commented:
Can you add a case to the ReaderTests/WriterTests that fails without this change?

The tests added in https://github.com/tmds/Tmds.DBus/pull/18 include an 8-item tuple, but they do not reveal the issue.

Umdrahschua commented:
I tried to add a test before this PR, but I had some issues with it... Yes, I will look into this and try to add a test.

I'm not sure who is creating this *nested tuple*, but a *nested tuple* with only one item makes no sense to me. In this case.


If you like to reproduce the issue, you have to install *timesyncd*, a replacement for chrony. It is part of systemd.

You can run `dbus-send --system --type=method_call --print-reply --dest=org.freedesktop.timesync1 /org/freedesktop/timesync1 org.freedesktop.DBus.Properties.GetAll string:org.freedesktop.timesync1.Manager` to get all properties from command line.

I created a branch with a test case that I used for debugging: https://github.com/Umdrahschua/Tmds.DBus/tree/test/timesync1

Umdrahschua commented:
@tmds You already add a test for this

https://github.com/tmds/Tmds.DBus/blob/a62998f9bf8de28a76747b6d1fe773ff013f657a/test/Tmds.DBus.Tests/ReaderTests.cs#L438

We have here a tuple with 8 items and if I stop the debugger on the right place, then we see the a nested tuple

```
{System.ValueTuple`8[System.Byte,System.Byte,System.Byte,System.Byte,System.Byte,System.Byte,System.Byte,System.ValueTuple`1[System.Byte]]}
```

And the nested tuple is not detected as tuple. But the test is running fine and does not read over the end. So, any idea to trigger the issue in the test? 

(of course, I do use original code and not the fixed code. I'd like to see the issue as first)

tmds commented:
> So, any idea to trigger the issue in the test?

I think the issue is triggered by padding.

With `((byte)1, (byte)2, (byte)3, (byte)4, (byte)5, (byte)6, (byte)7, (byte)8)`, the last byte is at 8-byte alignment.

May be we can reproduce the issue if we move it up 1 byte by changing one the first element into a short: `((short)1, (byte)2, (byte)3, (byte)4, (byte)5, (byte)6, (byte)7, (byte)8)`?

Umdrahschua commented:
Still try to write a test for this... but I do not found a way to trigger the issue. 

Now, I try change the `PersonProperties`-test, because I'd like run into method `ReadDictionaryObject`. But I have so my problems with the padding...

Umdrahschua commented:
@tmds 

I updated the reader tests. Before the fix, the test was failing (`_pos != endPos`). After the fix, it works.

I also updated the writer tests.

As mentioned in the issue, I also changed the [DirectCopy() call](https://github.com/tmds/Tmds.DBus/pull/216/commits/fb4fba30b8a3db5c64712bcb82eb7e746fc9b2d7). But if I revert the [first change](https://github.com/tmds/Tmds.DBus/pull/216/commits/eaa0eab4f6187fb7640d1ea206b4fd20518d030b), then the tests do not work. But getting-all-properties-from-timesync1 does work. I do not know why this happens, but rewriting the tuple-detection fixed my issue.

tmds commented on src/Tmds.DBus/Protocol/MessageReader.cs:
```diff
@@ -599,7 +599,7 @@ object MarshalStruct (Type structType, FieldInfo[] fis)
             int sof = Marshal.SizeOf (fis[0].FieldType);
             GCHandle handle = GCHandle.Alloc (strct, GCHandleType.Pinned);
-            DirectCopy (sof, (uint)(fis.Length * sof), handle);
+            DirectCopy (sof, (uint)fis.Length, handle);
```
We need to make all the callers either pass the field length, or the byte length.

Suggestion:
```diff
diff --git a/src/Tmds.DBus/Protocol/MessageReader.cs b/src/Tmds.DBus/Protocol/MessageReader.cs
index f1509a9..b8ee33e 100644
--- a/src/Tmds.DBus/Protocol/MessageReader.cs
+++ b/src/Tmds.DBus/Protocol/MessageReader.cs
@@ -504,30 +504,30 @@ TArray[] MarshalArray<TArray> (uint length)
                 _pos += (int)length;
             } else {
                 GCHandle handle = GCHandle.Alloc (array, GCHandleType.Pinned);
-                DirectCopy (sof, length, handle);
+                DirectCopy (sof, array.Length, handle);
                 handle.Free ();
             }
 
             return array;
         }
 
-        void DirectCopy (int sof, uint length, GCHandle handle)
+        void DirectCopy (int sof, int length, GCHandle handle)
         {
             DirectCopy (sof, length, handle.AddrOfPinnedObject ());
         }
 
-        unsafe void DirectCopy (int sof, uint length, IntPtr handle)
+        unsafe void DirectCopy (int sof, int length, IntPtr handle)
         {
+            int byteLength = length * sof;
             if (_endianness == Environment.NativeEndianness) {
-                Marshal.Copy (_data.Array, _data.Offset + _pos, handle, (int)length);
+                Marshal.Copy (_data.Array, _data.Offset + _pos, handle, byteLength);
             } else {
                 byte* ptr = (byte*)(void*)handle;
-                for (int i = _pos; i < _pos + length; i += sof)
+                for (int i = _pos; i < _pos + byteLength; i += sof)
                     for (int j = i; j < i + sof; j++)
                         ptr[2 * i - _pos + (sof - 1) - j] = _data.Array[_data.Offset + j];
             }
-
-            _pos += (int)length * sof;
+            _pos += byteLength;
         }
 
         bool[] MarshalBoolArray (uint length)
@@ -599,7 +599,7 @@ object MarshalStruct (Type structType, FieldInfo[] fis)
             object strct = Activator.CreateInstance (structType);
             int sof = Marshal.SizeOf (fis[0].FieldType);
             GCHandle handle = GCHandle.Alloc (strct, GCHandleType.Pinned);
-            DirectCopy (sof, (uint)(fis.Length * sof), handle);
+            DirectCopy (sof, fis.Length, handle);
             handle.Free ();
 
             return strct;
```

Umdrahschua replied:
For me is this Okay. Should I do the changes or you?

tmds replied:
Please include them in the PR.

tmds commented:
> As mentioned in the issue, I also changed the [DirectCopy() call](https://github.com/tmds/Tmds.DBus/pull/216/commits/fb4fba30b8a3db5c64712bcb82eb7e746fc9b2d7). But if I revert the [first change](https://github.com/tmds/Tmds.DBus/pull/216/commits/eaa0eab4f6187fb7640d1ea206b4fd20518d030b), then the tests do not work. But getting-all-properties-from-timesync1 does work. I do not know why this happens, but rewriting the tuple-detection fixed my issue.

It's fixing two bugs. One is a padding bug for structs. The other is the offset that is wrong in `DirectCopy`.

Thank you for debugging this, and making a PR to fix it, with test cases!

Umdrahschua commented:
> Thank you for debugging this, and making a PR to fix it, with test cases!

No problem.

PR is up-to-date now.

tmds commented:
This is part of 0.16.0 which was just published to nuget.org.
